internal/events/batch_pin_complete.go

```diff
@@ -68,8 +69,13 @@ func (em *eventManager) BatchPinComplete(bi blockchain.Plugin, batchPin *blockch
 			if err := em.persistContexts(ctx, batchPin, signingKey, private); err != nil {
 				return err
 			}
-			// Kick off a download for broadcast batches
-			if !private {
+
+			batch, err := em.database.GetBatchByID(ctx, batchPin.BatchID)
+			if err != nil {
+				return err
+			}
+			// Kick off a download for broadcast batches if the batch isn't already persisted
+			if !private && batch == nil {
```

Comment:
Alright, so the issue was not actually that the sender's aggregator was incorrectly receiving and confirming the pin before the batch was persisted - the issue was that the batch was being re-persisted because we were downloading the batch from IPFS as the sender which was incorrect.

The sender persists the batch before uploading to IPFS, so theres no need for it to unnecessarily download from IPFS and then re-persist the batch. So after ensuring the pins are persisted when receiving the `BatchPinComplete` event from the chain, we now do a check to see if the batch is already persisted which avoids the issue.

The reason why this didn't affect other batch / messages sent after reigsteration was because of the `sentByUs` function being used in `persist_batch.go` which used the `Node` of the batch to check if the node authored it. For registration batches, the node hasn't been registered so that didn't protect us.

The extra DB lookup prevents wasting a download worker for downloading batches when we are the sender, as well as if the batch pin events from etheconnect are ever replayed (i.e. the eventstream is deleted) which are both added benefits with this fix.

Only question remains: is this the most appropriate way to look up if the batch is already persisted? Should it be going through the operator manager or checking the cache first? @peterbroadhurst 

Want to make sure this small bit of code is correct before updating the unit tests for it.

Reply:
Great investigation 👍 

We have a batch cache, that should be used here rather than going directly to the DB.
See `GetBatchForPin`.

One potential thing I've been thinking through here - we have considered the database design of FireFly such that it should be pluggable to document databases without SQL style transactions in the future.
I was digging around to think about if there's a possibility that because we write the batch (which just has the manifest) to the DB before all the contents whether there's a possibility for an issue when one of those DBs come along. Basically whether we should swap the order of `validateAndPersistBatchContent` and `UpsertBatch` in `persistBatch` for future proofing to when those DBs come along.

Reply:
e.g. do this line first:
https://github.com/hyperledger/firefly/blob/41a5a155121aaeeb076a245399f3cd1e397ba429/internal/events/persist_batch.go#L82
Then this line second:
https://github.com/hyperledger/firefly/blob/41a5a155121aaeeb076a245399f3cd1e397ba429/internal/events/persist_batch.go#L72

Reply:
> We have a batch cache, that I think should be used here rather than going directly to the DB.
See GetBatchForPin.

Makes sense, only concern I have is if somehow the batch is not in the cache but its been persisted when receiving the pin event. Should it be check cache, if not in cache check DB?

> Basically whether we should swap the order of validateAndPersistBatchContent and UpsertBatch in persistBatch for future proofing to when those DBs come along.

Sounds reasonable to me! Will go ahead and add that in as part of this.

Reply:
Saw that `GetBatchForPin` falls through the DB in the event of an empty cache so this was exactly was what's needed and is working.